### ISSUE TITLE
Plugin: Backport PHP changes for WordPress 6.1 release

### ISSUE THREAD

**Issue by ockham** [WordPress/gutenberg#43440]
## Description

Related [Trac ticket 56467](https://core.trac.wordpress.org/ticket/56467).

This is the tracking issue to coordinate the process of backporting all PHP changes added in the Gutenberg plugin that needs to be backported for the WordPress 6.1 release.

### [`lib/block-supports`](https://github.com/WordPress/gutenberg/tree/trunk/lib/block-supports)

- [x] ~[`border.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/border.php)~ https://github.com/WordPress/wordpress-develop/pull/3204 🟢
- [x] ~[`colors.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php)~ https://github.com/WordPress/wordpress-develop/pull/3204 🟢
- [x] ~[`dimensions.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/dimensions.php)~ _No changes since WP 6.0._
- [x] ~[`duotone.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/duotone.php)~ https://github.com/WordPress/wordpress-develop/pull/3210 🟢
- [x] ~[`elements.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/dimensions.php)~ https://github.com/WordPress/wordpress-develop/pull/3204 🟢
- [x] ~[`layout.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php)~ @andrewserong ([ref](https://github.com/WordPress/gutenberg/issues/43440#issuecomment-1232172697)). _Blocked by https://github.com/WordPress/wordpress-develop/pull/3212 and https://github.com/WordPress/wordpress-develop/pull/2928._ https://github.com/WordPress/wordpress-develop/pull/3199 https://github.com/WordPress/wordpress-develop/pull/3254 🟢
- [x] ~[`spacing.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/spacing.php)~ https://github.com/WordPress/wordpress-develop/pull/3204 🟢
- [x] ~[`typography.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/typography.php)~ https://github.com/WordPress/wordpress-develop/pull/3203 🟢
- [x] ~[`utils.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/utils.php)~ _No changes since WP 6.0._

### [`lib/compat/wordpress-6.1`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1)

- [x] ~[`block-editor-settings.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/block-editor-settings.php)~ - Layout: [#42906](https://github.com/WordPress/gutenberg/pull/42906) Spacing: [#42173](https://github.com/WordPress/gutenberg/pull/42173) 🟢
- [ ] [`block-patterns.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/block-patterns.php) - 🔴 Patterns registered in this file are not in Core.
- [x] ~[`block-template-utils.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/block-template-utils.php)~ Site editor: [#3214](https://github.com/WordPress/wordpress-develop/pull/3214) 🟢
- [ ] [`blocks.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/blocks.php) 🔴 It seems that the multiple  `viewScripts` feature has not been backported.
- [x] ~[`class-gutenberg-rest-block-patterns-controller.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php)~  No changes since WP 6.0.
- [x] ~[`class-gutenberg-rest-templates-controller.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php)~ No changes since WP 6.0.
- [x] ~[`class-wp-theme-json-6-1.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php)~ https://github.com/WordPress/wordpress-develop/pull/3206 https://github.com/WordPress/wordpress-develop/pull/3205 https://github.com/WordPress/wordpress-develop/pull/3255 🟢
- [x] ~[`class-wp-theme-json-resolver-6-1.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php)~ https://github.com/WordPress/wordpress-develop/pull/3255 ⚠️ Core has more 6.1 functions than the Gutenberg file.
- [x] ~[`date-settings.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/date-settings.php)~ Code is already in Core 🟢
- [x] ~[`edit-form-blocks.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/edit-form-blocks.php)~ https://github.com/WordPress/wordpress-develop/pull/3156 @Mamaduka 🟢
- [x] ~[`get-global-styles-and-settings.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/get-global-styles-and-settings.php)~ 🟢
- [x] ~[`persisted-preferences.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/persisted-preferences.php)~ https://github.com/WordPress/wordpress-develop/pull/3219 🟢
- [x] ~[`rest-api.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/rest-api.php)~  🟢
- [x] ~[`script-loader.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/script-loader.php)~ Style engine changes in [WordPress/wordpress-develop#3218](https://github.com/WordPress/wordpress-develop/pull/3218) 🟢
- [ ] [`theme-i18n.json`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/theme-i18n.json) 🔴
- [x] ~[`theme.json`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/theme.json)~ https://github.com/WordPress/wordpress-develop/pull/3205 🟢
- [x] ~[`theme.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/theme.php)~ @andrewserong _Should be covered by layout related work. Only affected by https://github.com/WordPress/gutenberg/pull/42544._ 🟢
- [x] ~[`wp-theme-get-post-templates.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/wp-theme-get-post-templates.php)~ _(No need to backport, see https://github.com/WordPress/gutenberg/issues/43440#issuecomment-1228285020.)_

### [`lib/experimental`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental)

- [x] ~[`block-editor-settings-mobile.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/block-editor-settings-mobile.php)~ _No need to backport: Only used for mobile editor. @geriux confirmed that the endpoint is only available from the GB plugin for now._
- [x] ~[`blocks.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/blocks.php)~ _Deleted._
- [ ] [`class-wp-rest-block-editor-settings-controller.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-rest-block-editor-settings-controller.php)
- [ ] [`class-wp-rest-customizer-nonces.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-rest-customizer-nonces.php)
- [x] ~[`class-wp-theme-json-gutenberg.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-theme-json-gutenberg.php)~ _No need to backport -- "alias" wrapper around `WP_Theme_JSON_6_1`._
- [ ] [`class-wp-theme-json-resolver-gutenberg.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-theme-json-resolver-gutenberg.php)
- ~[ ] [`class-wp-webfonts-provider-local.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-webfonts-provider-local.php)~ 🔴  Webfonts API is still in development and not ready for WP 6.1.
- ~[ ] [`class-wp-webfonts-provider.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-webfonts-provider.php)~ 🔴  Webfonts API is still in development and not ready for WP 6.1.
- ~[ ] [`class-wp-webfonts.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/class-wp-webfonts.php)~ 🔴  Webfonts API is still in development and not ready for WP 6.1.
- [ ] [`editor-settings.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/editor-settings.php)
- [ ] [`navigation-page.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/navigation-page.php)
- [ ] [`navigation-theme-opt-in.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/navigation-theme-opt-in.php)
- ~[ ] [`register-webfonts-from-theme-json.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/register-webfonts-from-theme-json.php)~ 🔴  Webfonts API is still in development and not ready for WP 6.1.
- [ ] [`rest-api.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/rest-api.php)
- ~[ ] [`webfonts.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/webfonts.php)~ 🔴  Webfonts API is still in development and not ready for WP 6.1.

**Legend:**

🟢 – confirmed for WP 6.1
🟡 – needs decision
🔴 – moved to future WP release

## Getting Involved

If you are interested in helping with the effort, can you comment (or edit the issue's description) with your name next to the file if you plan to help with backporting. It would also help to link to Track issues / GitHub PRs in WordPress core when they are available.

Action items when working on backports:
- Update the files to follow [best practices](https://github.com/WordPress/gutenberg/tree/trunk/lib#best-practices). Mostly rename functions/classes to use `wp_` prefix instead of `gutenberg_` and guard with declaration checks for code that needs to be backported to WordPress core.
- Identify files and functionality that need to be backported to WordPress core.
- Ensure all files, classes, methods, properties, and functions have `@since 6.1.0` documentation tag included.
- Ensure all functionality backported to WordPress core has unit tests written.

Note that @hellofromtonya has started work to smoothen the backporting process, by adjusting Gutenberg's test and lint structure to more closely resemble Core's (to avoid having to fix code in backport PRs when they don't pass `wordpress-develop`'s tests even though they did pass Gutenberg's): #43333

---

_Issue modeled after #39889._

**Comment by Mamaduka** [comment]
I'll create a patch for changes in [`edit-form-blocks.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/edit-form-blocks.php) file.

The `gutenberg_add_site_icon_url_to_index` feature backport from `rest-api.php` depends if we merge #43514 or not.

@ntsekouras, correct me if I'm wrong, but the [`wp-theme-get-post-templates.php`](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/wp-theme-get-post-templates.php) doesn't need to be ported into the core.

**Comment by Aljullu** [comment]
Heads-up that I opened a PR in wordpress-develop: https://github.com/WordPress/wordpress-develop/pull/3131 which implements the same fix as #43597 in `block-template-utils.php`.

Related track issue: https://core.trac.wordpress.org/ticket/55881#ticket

**Comment by ntsekouras** [comment]
>@ntsekouras, correct me if I'm wrong, but the [wp-theme-get-post-templates.php](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/wp-theme-get-post-templates.php) doesn't need to be ported into the core.

Good catch @Mamaduka ! It seems so.

**Comment by ockham** [comment]
Apologies for the mass ping -- I'd like to get the backports work off the ground 😊 

_tl;dr: If you've added PHP code to Gutenberg's `lib/` directory since the WP 6.0 release, it will need backporting to [`wordpress-develop`](https://github.com/WordPress/wordpress-develop) in order to be included in WP 6.1. Backporting is typically easiest for the people who originally implemented the relevant feature. Please refer to the list in the issue description and leave a comment if you're willing to backport code. If you're unfamiliar with the process of backporting, don't hesitate to ping me (@ockham) or @michalczaplinski -- we're happy to help!_

PRs that are strickenthrough are already merged in WordPress Core

### [lib/block-supports/](https://github.com/WordPress/gutenberg/tree/trunk/lib/block-supports/)

- [border.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/border.php)
  - #42719
  - #42140
  - #42113
  - ~~#41527~~
  - #40531
  - #37770
@ramonjd @glendaviesnz @aaronrobertshaw
- [colors.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/colors.php)
  - #42719
  - #42140
  - #42113
  - #40332
@ramonjd
- [dimensions.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/dimensions.php)

- [duotone.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/duotone.php)
  - ~~#42020~~
  - ~~#39564~~
@jorgefilipecosta @ajlende
- [elements.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/elements.php)
  - #43103
  - #42878
  - #42452
  - #42719
  - #42140
  - #42113
  - #40987
  - #40594
  - ~~#37728~~
@ramonjd @oandregal @scruffian
- [layout.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/layout.php)
  - #43103
  - #43052
  - #42970
  - #42939
  - #42878
  - #42452
  - ~~#42544~~
  - #42719
  - ~~#42085~~
  - ~~#40875~~
  - #41306
  - #41885
  - #41487
  - #41098
  - #41018
@ramonjd @aristath @matiasbenedetto @andrewserong @tellthemachines @youknowriad @glendaviesnz
- [spacing.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/spacing.php)
  - #42719
  - #42140
  - #42113
  - ~~#41527~~
  - #40332
@ramonjd @glendaviesnz
- [typography.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/typography.php)
  - #42757
  - #42719
  - #42688
  - ~~#39529~~
  - #42140
  - #42113
  - #40332
@ramonjd
- [utils.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/utils.php)

### [lib/compat/wordpress-6.1/](https://github.com/WordPress/gutenberg/tree/trunk/lib/compat/wordpress-6.1/)

- [block-editor-settings.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-editor-settings.php)
  - #42173
  - #42906
  - #42818
  - ~~#42544~~
  - ~~#40875~~
  - ~~#41527~~
  - #40818
  - ~~#40185~~
@glendaviesnz @andrewserong @Soean @jsnajdr @oandregal
- [block-patterns.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-patterns.php)
  - #41791
@jorgefilipecosta
- [block-template-utils.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-template-utils.php)
  - #42520
  - #42586
  - #41387
  - ~~#41302~~
@ntsekouras
- [blocks.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/blocks.php)
  - #43004
  - ~~#40875~~
  - #40136
  - #40349
@andrewserong @aristath @gziolo
- [class-gutenberg-rest-block-patterns-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php)
  - #41791
  - #40902
  - #40416
  - #40259
  - #40265
  - #40254
  - ~#40252~ _(This is a "back-backport" from Core.)_
@jorgefilipecosta @anton-vlasenko @Mamaduka @ntsekouras @gziolo @jsnajdr
- [class-gutenberg-rest-templates-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php)
  - #42520
  - #41189
  - #41387
@ntsekouras
- [class-wp-theme-json-6-1.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php)
  - #43167
  - ~~#43105~~
  - #43088
  - ~~#43101~~
  - #42776
  - ~~#42544~~
  - ~~#42665~~
  - #42669
  - ~~#39529~~
  - ~~#42085~~
  - ~~#42087~~
  - ~~#40875~~
  - ~~#41981~~
  - #42096
  - #42072
  - ~~#41527~~
  - ~~#41140~~
  - #41696
  - #41786
  - #41822
  - #41753
  - ~~#41335~~
  - ~~#41240~~
  - ~~#41217~~
  - #40260
  - ~~#41160~~
  - #40889
@Mamaduka @glendaviesnz @MaggieCabrera @scruffian @andrewserong @ramonjd @tellthemachines @aaronrobertshaw @getdave @matiasbenedetto @torounit
- [class-wp-theme-json-resolver-6-1.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php)
  - #41306
  - ~~#41527~~
  - #41822
@youknowriad @glendaviesnz @scruffian
- [date-settings.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/date-settings.php)
  - #41648
@noisysocks
- [edit-form-blocks.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/edit-form-blocks.php)
  - #42413
  - ~~#42209~~
@Mamaduka
- [get-global-styles-and-settings.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/get-global-styles-and-settings.php)
  - ~~#40875~~
  - #42005
  - #41446
  - ~~#41160~~
@andrewserong @scruffian @glendaviesnz
- [persisted-preferences.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/persisted-preferences.php)
  - #40637
  - #39795
@talldan
- [rest-api.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/rest-api.php)
  - ~~#42957~~
  - #41791
  - #41189
  - #41387
@Mamaduka @jorgefilipecosta @ntsekouras
- [script-loader.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/script-loader.php)
  - #42880
  - #43073
  - #42207
  - #42005
  - #41015
  - ~~#41160~~
@ramonjd @scruffian @glendaviesnz
- [theme-i18n.json](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/theme-i18n.json)
  - ~~#41527~~
  - #40994
@glendaviesnz @gziolo
- [theme.json](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/theme.json)
  - #42173
  - #42053
  - ~~#42085~~
  - ~~#40875~~
  - #42012
  - ~~#41527~~
  - #41934
  - #41786
  - #41822
@glendaviesnz @matiasbenedetto @tellthemachines @andrewserong @scruffian @getdave
- [theme.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/theme.php)
  - ~~#42544~~
@andrewserong

### [lib/experimental/](https://github.com/WordPress/gutenberg/tree/trunk/lib/experimental/)

- [block-editor-settings-mobile.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/block-editor-settings-mobile.php)
  - #42697
  - #25892
@ellatrix @oandregal
- [blocks.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/blocks.php)
  - #42697
  - #25892
@ellatrix @oandregal
- [class-wp-rest-block-editor-settings-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/class-wp-rest-block-editor-settings-controller.php)
  - #42173
  - ~~#42544~~
  - #42697
  - ~~#41527~~
@glendaviesnz @andrewserong @ellatrix
- [class-wp-rest-customizer-nonces.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/class-wp-rest-customizer-nonces.php)

- [class-wp-theme-json-gutenberg.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/class-wp-theme-json-gutenberg.php)
  - #42776
  - #40889
@scruffian
- [class-wp-theme-json-resolver-gutenberg.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/experimental/class-wp-theme-json-resolver-gutenberg.php)
  - #42756
  - #43337

---

_List modeled after https://github.com/WordPress/gutenberg/issues/39889#issuecomment-1087604073. Created using the script from #43360._

**Comment by glendaviesnz** [comment]
I can take care of all the spacing size preset related changes.

**Comment by Mamaduka** [comment]
@ockham, @michalczaplinski, do we have a core tracking ticket to reference in Core PRs?

**Comment by ockham** [comment]
> @ockham, @michalczaplinski, do we have a core tracking ticket to reference in Core PRs?

@Mamaduka Ah, we didn't -- my apologies! I've filed one now: https://core.trac.wordpress.org/ticket/56467

**Comment by andrewserong** [comment]
I'm happy to help out with Layout backports. I'm travelling at the moment but can start looking at some of them next week when I'm back. Some of them might have a couple of challenges — layout.php currently depends on the style engine work in the style engine package (so just CCing @ramonjd there), which might need to land first?

Also, I was wondering how we deal with PHP changes that require JS changes to be shipped at the same time? I _think_ some of the Layout work might need to happen all at once 🤔. There's definitely pieces that can be done in separate PRs first, though, so I'll give it a bit more thought to see what sort of order might work.

Just linking a couple of other pre-requisites for Layout in Gutenberg being merged to core:

* Trac ticket [55966](https://core.trac.wordpress.org/ticket/55966), which adds in support for additional CSS functions to `safecss_filter_attr` (like `clamp`)
* Core PR (https://github.com/WordPress/wordpress-develop/pull/2928) to add in layout related CSS properties to `safe_style_css_filter`

I believe the above two tickets/PRs will need to land prior to the `layout.php` changes being backported.

**Comment by Mamaduka** [comment]
@ockham,

My backport PRs are ready:

* https://github.com/WordPress/wordpress-develop/pull/3156
* https://github.com/WordPress/wordpress-develop/pull/3157

The https://github.com/WordPress/gutenberg/pull/40416 should already be in Core, cherry-picked for a minor release - https://github.com/WordPress/wordpress-develop/pull/2600.

I've only done a minor code quality work for `class-wp-theme-json-6-1.php`, but happy to help as needed.

**Comment by ockham** [comment]
Small update: In https://github.com/WordPress/wordpress-develop/pull/3154, I'm attempting a package sync to update Core to the latest package versions -- mostly as a dry-run for Beta 1. (I would like to actually merge it once its ready -- see its PR desc for context).

In order to unblock that PR, I've filed a small [backport PR](https://github.com/WordPress/wordpress-develop/pull/3158) to add the `$priority` param to `wp_enqueue_block_support_styles`.

**Comment by gziolo** [comment]
> In order to unblock that PR, I've filed a small [backport PR](https://github.com/WordPress/wordpress-develop/pull/3158) to add the `$priority` param to `wp_enqueue_block_support_styles`.

In addition to that, we need to refactor all occurrences of `WP_Theme_JSON_Gutenberg` and `gutenberg_enqueue_block_support_styles` in core blocks in the Gutenberg plugin. That might not be trivial as the plugin still supports WordPress 6.0 and 6.1 where some functionality is missing.

**Comment by ockham** [comment]
> > In order to unblock that PR, I've filed a small [backport PR](https://github.com/WordPress/wordpress-develop/pull/3158) to add the `$priority` param to `wp_enqueue_block_support_styles`.
> 
> In addition to that, we need to refactor all occurrences of `WP_Theme_JSON_Gutenberg` and `gutenberg_enqueue_block_support_styles` in core blocks in the Gutenberg plugin. That might not be trivial as the plugin still supports WordPress 6.0 and 6.1 where some functionality is missing.

Good point. Here's one approach for `gutenberg_enqueue_block_support_styles`: https://github.com/WordPress/gutenberg/pull/43779.

**Comment by noisysocks** [comment]
Thanks for the nice list—that's a cool script! I've backported https://github.com/WordPress/gutenberg/pull/41648 (in [r54079](https://core.trac.wordpress.org/changeset/54079)) and checked it off the list in the description.

**Comment by ramonjd** [comment]
I'm working on prepping [packages/style-engine](https://github.com/WordPress/gutenberg/tree/trunk/packages/style-engine) for inclusion in core. 

WIP here: https://github.com/WordPress/wordpress-develop/pull/3199

This will unblock most of the lib/block-supports migrations given that they're using style engine methods.

It requires https://core.trac.wordpress.org/ticket/55966, otherwise I'll need to migrate [gutenberg_safecss_filter_attr_allow_css_6_1](https://github.com/WordPress/gutenberg/commit/60adc79292c4a8c41ac206152ace4f2dd5faed8f#diff-4afd960d3bcdfb07993478077c20aac762bece548a901eafc7f93b4abae08023R39) to core to make the Style Engine work. 😱

**Comment by ockham** [comment]
> It requires https://core.trac.wordpress.org/ticket/55966, otherwise I'll need to migrate [gutenberg_safecss_filter_attr_allow_css_6_1](https://github.com/WordPress/gutenberg/commit/60adc79292c4a8c41ac206152ace4f2dd5faed8f#diff-4afd960d3bcdfb07993478077c20aac762bece548a901eafc7f93b4abae08023R39) to core to make the Style Engine work. 😱

I've [asked in #core](https://wordpress.slack.com/archives/C02RQBWTW/p1662461774015659) for someone to merge 55966's patch 🙂

**Comment by ockham** [comment]
As the Feature Freeze is only two weeks away, I'll start pinging some folks who I know have good domain knowledge of certain areas and can help backporting some files 😊 

@ntsekouras -- Could you file a backport PR for the changes you made to [`block-template-utils.php`](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-template-utils.php) and [`class-gutenberg-rest-templates-controller.php`](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php)? 🙏

**Comment by ockham** [comment]
> As the Feature Freeze is only two weeks away, I'll start pinging some folks who I know have good domain knowledge of certain areas and can help backporting some files 😊

@jorgefilipecosta -- Could you maybe file a backport PR for https://github.com/WordPress/gutenberg/pull/41791 (affecting [block-patterns.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-patterns.php), [class-gutenberg-rest-block-patterns-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php), and [rest-api.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/rest-api.php))? 🙏

**Comment by jorgefilipecosta** [comment]
> > As the Feature Freeze is only two weeks away, I'll start pinging some folks who I know have good domain knowledge of certain areas and can help backporting some files 😊
> 
> @jorgefilipecosta -- Could you maybe file a backport PR for #41791 (affecting [block-patterns.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-patterns.php), [class-gutenberg-rest-block-patterns-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php), and [rest-api.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/rest-api.php))? 🙏

Sure 👍

**Comment by ockham** [comment]
>As the Feature Freeze is only two weeks away, I'll start pinging some folks who I know have good domain knowledge of certain areas and can help backporting some files 😊

@oandregal -- Could you maybe file a backport PR for #25892 (Quote v2)? 🙏

**Comment by talldan** [comment]
I'm taking a look at persisted preferences.

**Comment by MaggieCabrera** [comment]
I'm working on backporting the changes and additions to the Elements API [here](https://github.com/WordPress/wordpress-develop/pull/3206). The PR has a list of the PRs I'm backporting, the `class-wp-theme-json-6-1.php` file has other changes unrelated to elements that I won't be adding to said PR, this is the list of the PRs it's NOT going to include:


https://github.com/WordPress/gutenberg/pull/41217
https://github.com/WordPress/gutenberg/pull/41335
https://github.com/WordPress/gutenberg/pull/41527
https://github.com/WordPress/gutenberg/pull/40875
https://github.com/WordPress/gutenberg/pull/42087
https://github.com/WordPress/gutenberg/pull/42085
https://github.com/WordPress/gutenberg/pull/39529
https://github.com/WordPress/gutenberg/pull/42665
https://github.com/WordPress/gutenberg/pull/42544
https://github.com/WordPress/gutenberg/pull/43101
https://github.com/WordPress/gutenberg/pull/43105

**Comment by ajlende** [comment]
> - [duotone.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/block-supports/duotone.php)
>    - https://github.com/WordPress/gutenberg/pull/42020
>    - https://github.com/WordPress/gutenberg/pull/39564
> - [class-wp-theme-json-6-1.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php)
>   - https://github.com/WordPress/gutenberg/pull/41335

I'm working on the duotone backports

EDIT: duotone.php is done in https://github.com/WordPress/wordpress-develop/pull/3210 and I've talked with @MaggieCabrera for the class-wp-theme-json-6-1.php changes because they depend on the [Elements API backports](https://github.com/WordPress/wordpress-develop/pull/3206).

**Comment by oandregal** [comment]
:wave: sorry for the late response. I've looked at the pings but nothing needs backporting on my side:

- `elements.php`
  - #37728 this has been [backported for 6.0](https://github.com/WordPress/gutenberg/pull/37728#issuecomment-1108686176)
- `block-editor-settings.php`
  - https://github.com/WordPress/gutenberg/pull/40185 it [has been backported](https://github.com/WordPress/gutenberg/pull/40185#issuecomment-1102413528) for 6.0
- `lib/experimental/block-editor-settings-mobile.php#L18`
  - we should not backport this until mobile confirms the endpoint goes to core, see [context](https://github.com/WordPress/gutenberg/pull/39030/files#r814083346). cc @geriux 
- the bits for list v2 have been removed

**Comment by ntsekouras** [comment]
>@ntsekouras -- Could you file a backport PR for the changes you made to [block-template-utils.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/block-template-utils.php) and [class-gutenberg-rest-templates-controller.php](https://github.com/WordPress/gutenberg/commits/trunk/lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php)? 🙏

I'll look at it, but we probably need to wait @jorgefilipecosta for a PR that updates some related logic there.

**Comment by ntsekouras** [comment]
> - https://github.com/WordPress/gutenberg/pull/41302
@ntsekouras

PR: https://github.com/WordPress/wordpress-develop/pull/3214

**Comment by cbravobernal** [comment]
I will take a look at:
- https://github.com/WordPress/gutenberg/pull/41981
- https://github.com/WordPress/gutenberg/pull/41140
- https://github.com/WordPress/gutenberg/pull/41240

Backported in https://github.com/WordPress/wordpress-develop/pull/3215

**Comment by ramonjd** [comment]
Heads up that I think https://github.com/WordPress/wordpress-develop/pull/3199 (Style engine backport) is baked and ready to land. Thanks to @costdev for helping to prepare.

**Comment by ramonjd** [comment]
[This wordpress-develop PR](https://github.com/WordPress/wordpress-develop/pull/3218) backports the following changes:

- https://github.com/WordPress/gutenberg/pull/42880

**Comment by talldan** [comment]
PR here for persisted preferences - https://github.com/WordPress/wordpress-develop/pull/3219.

**Comment by cbravobernal** [comment]
> I believe the above two tickets/PRs will need to land prior to the `layout.php` changes being backported.

@andrewserong , those PR seems to be already merged. I guess we can keep going with layout.php. Thanks!!

**Comment by ntsekouras** [comment]
https://github.com/WordPress/wordpress-develop/pull/3221

Backports changes from: 
-  https://github.com/WordPress/gutenberg/pull/41387
- https://github.com/WordPress/gutenberg/pull/41189
- https://github.com/WordPress/gutenberg/pull/42520
- ~https://github.com/WordPress/gutenberg/pull/43597 (from @Mamaduka )~
- https://github.com/WordPress/gutenberg/pull/42586

**Comment by Mamaduka** [comment]
@ntsekouras, I think #43597 is covered by https://github.com/WordPress/wordpress-develop/pull/3131.

**Comment by gziolo** [comment]
I have an update for the `lib/compat/wordpress-6.1/blocks.php` file:

- there is an open [Trac ticket](https://core.trac.wordpress.org/ticket/53148#comment:3) and [PR](https://github.com/WordPress/wordpress-develop/pull/3222) for backporting `gutenberg_block_type_metadata_render_template` to WordPress core
- there is also an open [PR](https://github.com/WordPress/wordpress-develop/pull/3108) (still in progress) to backport support for multiple view scripts to WordPress core (`gutenberg_block_type_metadata_multiple_view_scripts`)
- I'm not sure if we immediately need `wp_enqueue_block_view_script` in WordPress core, @aristath should know better why we need it, but it might depend on how we land support for multiple view scripts

**Comment by ntsekouras** [comment]
https://github.com/WordPress/wordpress-develop/pull/3223

Backports changes for `build_query_vars_from_query_block` from:
- https://github.com/WordPress/gutenberg/pull/43590
- https://github.com/WordPress/gutenberg/pull/40933

**Comment by andrewserong** [comment]
> @andrewserong , those PR seems to be already merged. I guess we can keep going with layout.php. Thanks!!

Thanks for the ping @c4rl0sbr4v0! There's one other change that needs to land before we can backport `layout.php` in its current form, which is this backport for the style engine: https://github.com/WordPress/wordpress-develop/pull/3199 — In the meantime, I've started a WIP backport PR for the foundations of the Layout support in https://github.com/WordPress/wordpress-develop/pull/3205 (it's a bit of a moving target at the moment given other PRs landing in the Gutenberg repo prior to the feature freeze). My aim is to try to get `3205` mergeable without changes to `layout.php` and then do a follow-up PR for the `layout.php` changes 🤞

**Comment by andrewserong** [comment]
Update: I believe the first part of the Layout support backports (the foundational changes without `layout.php`) is ready for review in https://github.com/WordPress/wordpress-develop/pull/3205. It might need a look from folks who've also worked on the recent factor, so I've pinged @tellthemachines on the PR, too.

**Comment by anton-vlasenko** [comment]
> I'm not sure if we immediately need wp_enqueue_block_view_script in WordPress core, @aristath should know better why we need it, but it might depend on how we land support for multiple view scripts

FYI @gziolo @aristath
I've [submitted a PR](https://github.com/WordPress/gutenberg/pull/44100) to rename `wp_enqueue_block_view_script `to `gutenberg_enqueue_block_view_script` and deprecate the old name (`gutenberg_enqueue_block_view_script `).
IMO all PHP functions in the `lib/compat` folder should have the `gutenberg_` prefix to ease the backporting efforts and ensure good naming practices.

**Comment by noisysocks** [comment]
> IMO all PHP functions in the `lib/compat` folder should have the `gutenberg_` prefix to ease the backporting efforts and ensure good naming practices.

Very very open to a discussion about how to improve this (it's a mess!) but just noting that currently our stated approach is to use `wp_` with a `! function_exists()` guard. This lets plugins and theme developers use the `wp_` function and have their code work when running in WordPress 5.9 + Gutenberg, WordPress 6.0 + Gutenberg, and vanilla WordPress 6.1. For non-public APIs a `gutenberg_` prefix should be fine but we've not made this distinction in the past.

https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md

**Comment by anton-vlasenko** [comment]
~~to rename wp_enqueue_block_view_script to gutenberg_enqueue_block_view_script and deprecate the old name (gutenberg_enqueue_block_view_script ).~~
I've [reverted](https://github.com/WordPress/gutenberg/pull/44100/commits/54da4412d486f1e7b4e3aa45cf9418a1f15fac94) this change so that the code is in [accordance with the guideline](https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md#prefer-the-wp-prefix). Thanks for the input, @noisysocks.

**Comment by Mamaduka** [comment]
Backport PR for #42729 is ready for testing - https://github.com/WordPress/wordpress-develop/pull/3234.

**Comment by ironprogrammer** [comment]
Submitted [#3237](https://github.com/WordPress/wordpress-develop/pull/3237) as refresh of [#3203](https://github.com/WordPress/wordpress-develop/pull/3203), primarily to resolve failing PHP unit tests once style engine is in place.

**Comment by oandregal** [comment]
Prepared https://github.com/WordPress/wordpress-develop/pull/3247 to backport some recent changes: https://github.com/WordPress/gutenberg/pull/44015 https://github.com/WordPress/gutenberg/pull/44109 and https://github.com/WordPress/gutenberg/pull/44159

**Comment by andrewserong** [comment]
Thanks for all the help landing part 1 of the Layout refactor! Update on Layout and Spacing presets follow-up backports:

* Spacing Presets backport is ready for review from @glendaviesnz: https://github.com/WordPress/wordpress-develop/pull/3255
* Spacing Presets is designed to work with the backport for border, color, elements, spacing, so ideally that one might land first: https://github.com/WordPress/wordpress-develop/pull/3204
* Layout backport part 2 is ready for review: https://github.com/WordPress/wordpress-develop/pull/3254
* Layout part 2 depends on Backport script loader to land first: https://github.com/WordPress/wordpress-develop/pull/3259

**Comment by oandregal** [comment]
Prepared https://github.com/WordPress/wordpress-develop/pull/3261 to backport https://github.com/WordPress/gutenberg/pull/44189 (which is approved but I'm waiting for the tests to pass and will backport to 14.1).

**Comment by andrewserong** [comment]
Just looking over the theme JSON file and saw that `box-shadow` support hadn't been backported yet. I've opened up a PR in https://github.com/WordPress/wordpress-develop/pull/3274, which should pull in the final bits for that.

**Comment by ramonjd** [comment]
#### Block supports scripts loading

Added https://github.com/WordPress/wordpress-develop/pull/3273 as an extension of https://github.com/WordPress/wordpress-develop/pull/3259 cc @ockham

**Comment by oandregal** [comment]
I've prepared https://github.com/WordPress/wordpress-develop/pull/3313 to backport the changes at https://github.com/WordPress/gutenberg/pull/44363

**Comment by gziolo** [comment]
> I have an update for the `lib/compat/wordpress-6.1/blocks.php` file:
> 
> * there is an open [Trac ticket](https://core.trac.wordpress.org/ticket/53148#comment:3) and [PR](https://github.com/WordPress/wordpress-develop/pull/3222) for backporting `gutenberg_block_type_metadata_render_template` to WordPress core
> * there is also an open [PR](https://github.com/WordPress/wordpress-develop/pull/3108) (still in progress) to backport support for multiple view scripts to WordPress core (`gutenberg_block_type_metadata_multiple_view_scripts`)

I can confirm that all the changes are present in WordPress core.

> * I'm not sure if we immediately need `wp_enqueue_block_view_script` in WordPress core, @aristath should know better why we need it, but it might depend on how we land support for multiple view scripts

We didn't include this new API in WP core in 6.1. PR that reflects that in the plugin is ready for review: https://github.com/WordPress/gutenberg/pull/44414.

**Comment by scruffian** [comment]
I think this one is missing from some of the backports above: https://github.com/WordPress/wordpress-develop/pull/3319

**Comment by hellofromtonya** [comment]
I removed all of the Webfonts API files from the backport list. Why? The API is being re-architected, ie still in development. It will not be ready for WP 6.1. 🔴 

Please leave these files in the `lib/experimental/` directory. Thanks.

**Comment by cbravobernal** [comment]
Now that the 6.1 release is out, I think we can close this issue.
